packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m

```diff
@@ -536,123 +526,159 @@ - (NSDictionary *)encodingMap
-    reject(@"E_FILESYSTEM_PERMISSIONS",
+    reject(@"ERR_FILESYSTEM_PERMISSIONS",
            [NSString stringWithFormat:@"File '%@' isn't writable.", fileUri],
            nil);
     return;
   }
+  
+  if (![self _checkHeadersDictionary:options[@"headers"]]) {
+    reject(@"ERR_FILESYSTEM_INVALID_HEADERS_DICTIONARY",
+           @"Invalid headers dictionary. Keys and values should be strings.",
+           nil);
+    return;
+  }
+  
   NSData *resumeData = data ? [[NSData alloc] initWithBase64EncodedString:data options:0] : nil;
   [self _downloadResumableCreateSessionWithUrl:url
-                                withScopedPath:path
-                                      withUUID:uuid
-                                   withOptions:options
-                                withResumeData:resumeData
-                                  withResolver:resolve
-                                  withRejecter:reject];
+                                       fileUrl:localUrl
+                                          uuid:uuid
+                                        optins:options
+                                    resumeData:resumeData
+                                       resolve:resolve
+                                        reject:reject];
 }
 
 UM_EXPORT_METHOD_AS(downloadResumablePauseAsync,
                     downloadResumablePauseAsyncWithUUID:(NSString *)uuid
                     resolver:(UMPromiseResolveBlock)resolve
                     rejecter:(UMPromiseRejectBlock)reject)
 {
-  EXDownloadResumable *downloadResumable = (EXDownloadResumable *)self.downloadObjects[uuid];
-  if (downloadResumable == nil) {
-    reject(@"E_UNABLE_TO_PAUSE",
+  NSURLSessionDownloadTask *task = [_resumableManager getTask:uuid];
+  if (!task) {
+    reject(@"ERR_UNABLE_TO_PAUSE",
            [NSString stringWithFormat:@"There is no download object with UUID: %@", uuid],
            nil);
-  } else {
-    [downloadResumable.session getTasksWithCompletionHandler:^(NSArray<NSURLSessionDataTask *> * _Nonnull dataTasks, NSArray<NSURLSessionUploadTask *> * _Nonnull uploadTasks, NSArray<NSURLSessionDownloadTask *> * _Nonnull downloadTasks) {
-      NSURLSessionDownloadTask *downloadTask = [downloadTasks firstObject];
-      if (downloadTask) {
-        [downloadTask cancelByProducingResumeData:^(NSData * _Nullable resumeData) {
-          resolve(@{ @"resumeData": UMNullIfNil([resumeData base64EncodedStringWithOptions:0]) });
-        }];
-      } else {
-        reject(@"E_UNABLE_TO_PAUSE",
-               @"There was an error producing resume data",
-               nil);
-      }
-    }];
+    return;
   }
+  
+  UM_WEAKIFY(self);
+  [task cancelByProducingResumeData:^(NSData * _Nullable resumeData) {
+    UM_ENSURE_STRONGIFY(self);
+    resolve(@{ @"resumeData": UMNullIfNil([resumeData base64EncodedStringWithOptions:0]) });
+    [self.sessionTaskDispatcher unregisterTaskDelegate:task]; // It'll also unregister task from the ResumableManager
```

Comment:
Wouldn't `cancelByProducingResumeData` trigger the `- (void)URLSession:task:didCompleteWithError:` callback which would remove the task from the dispatcher even so?